### PR TITLE
Add support for ballerina program exit hooks

### LINKED ISSUE #9832: Need a proper solution to keep task timer/appointment running
**Description:**
Currently we have to use runtime:sleep() to stop the tasks from exiting. Need a proper solution instead. Tasks should run till timer.stop() is called. 

**Steps to reproduce:**
https://ballerina.io/learn/by-example/task-timer.html

### PR DESCRIPTION
**Pull request by Kishanthan**
## Purpose
The issue with tasks/appointments is that there is no proper way to keep them running. To solve this, we have to make the BVM to wait for all tasks to complete. But there should be a way to notify those running tasks in the event of BVM exit/interruption. This can be supported using ballerina program exit hooks approach.

A ballerina exit hook is similar to the java shutdown hook concept, where the application developer can register hooks (in ballerina these hooks will be a function) with the runtime. The runtime will invoke these hooks/functions whenever it receives a graceful interrupt signal (eg : Ctrl+C).

So how would this solve the issue with tasks?

If there are any registered hooks, the main will not exit. So, if we want to keep the main from exiting after a task creation, we have to register a hook. If you want to properly stop the tasks and clean them before the BVM exits, you can register a exit hook and do the cleaning related stuffs with them. The BVM will make sure to invoke all the registered hooks before it terminates.

The main design decision here is that, if there are any registered hooks only, the BVM will wait on the interrupt signal. But if there are no hooks registered, the BVM will anyway exit at the end of the execution, regardless of any are tasks created. The assumption here is that, if we have created a task and not added a exit hook, then it is considered a daemon task, which means the main can exit at the end.

How to add an exit hook?

```ballerina
import ballerina/task;
import ballerina/io;
import ballerina/runtime;


public function main (string... args) {

    (function() returns error?) onTriggerFunction = triggerFn;
    (function(error)) onErrorFunction = cleanupFn;

    //job 1 runs every 2 seconds.
    task:Appointment appointment1 = new task:Appointment(onTriggerFunction, 
                                        onErrorFunction, 
                                        "0/2 * * * * ?");


    var exitHookFn = () => {
        io:println("[INFO] stopping");
        appointment1.cancel();
    };
                                        
    appointment1.schedule();

    runtime:addExitHook(exitHookFn);
}

function cleanupFn(error e) {
    io:println("[ERROR] cleanup failed");
}

function triggerFn() {
    io:println("[INFO] triggered");
}
```

Resolves #9832

### CODE REVIEW

File: tests/ballerina-integration-test-utils/src/main/java/org/ballerinalang/test/context/BServer.java [hunk 1]
```diff
@@ -117,6 +117,14 @@ void startServer(String sourceRoot, String packagePath, String[] args,
      */
     void shutdownServer() throws BallerinaTestException;
 
+
```
Comment (Shan1024) [inline tests/ballerina-integration-test-utils/src/main/java/org/ballerinalang/test/context/BServer.java]: Extra newline?

File: tests/ballerina-integration-test-utils/src/main/java/org/ballerinalang/test/context/BServerInstance.java [hunk 2]
```diff
@@ -231,6 +254,14 @@ private void cleanupServer() {
...
     }
 
+    private void gracefulCleanupServer() throws InterruptedException {
+        process.waitFor();
+        //wait until port to close
+        Utils.waitForPortsToClose(requiredPorts, 30000);
+        log.info("Server Stopped Successfully");
+    }
+
+
```
Comment (Shan1024) [inline tests/ballerina-integration-test-utils/src/main/java/org/ballerinalang/test/context/BServerInstance.java]: Extra newline?

File: tests/ballerina-integration-test/src/test/resources/runtime/exit_hook.bal [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/io;
+import ballerina/runtime;
+
+
```
Comment (Shan1024) [inline tests/ballerina-integration-test/src/test/resources/runtime/exit_hook.bal]: Extra newline?

File: tests/ballerina-integration-test/src/test/resources/runtime/exit_hook.bal [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                _ = caller -> respond(resp);
+            }
+            any | () => {
+                io:println("Error while fetching string payload");
+            }
+        }
+    }
+}
+
+
```
Comment (Shan1024) [inline tests/ballerina-integration-test/src/test/resources/runtime/exit_hook.bal]: Extra newline?

**Comment by Kishanthan** [comment]
Closing this as we have to come up with an alternative design approach in handling signals in ballerina. The related doc is - https://docs.google.com/document/d/16W36cEprx1j2pVzd6B0bOwQ-xA0mpt8nV4_LZPqHMjA/edit?usp=sharing

**Pull request closed without merging**